src/object/cli_shard.c

```diff
@@ -289,6 +289,107 @@ dc_rw_cb_csum_verify(const struct rw_cb_args *rw_args)
 	return rc;
 }
 
+static int
+iom_recx_merge(daos_iom_t *dst, daos_recx_t *recx)
+{
+	daos_recx_t	*tmpr;
+	uint32_t	 i;
+
+	for (i = 0; i < dst->iom_nr_out; i++) {
+		tmpr = &dst->iom_recxs[i];
+		if (DAOS_RECX_PTR_OVERLAP(tmpr, recx) ||
+		    DAOS_RECX_PTR_ADJACENT(tmpr, recx)) {
+			daos_recx_merge(recx, tmpr);
+			return 0;
+		}
+	}
+
+	if (dst->iom_nr_out < dst->iom_nr) {
+		dst->iom_recxs[dst->iom_nr_out] = *recx;
+		dst->iom_nr_out++;
+		return 0;
+	}
+
+	return -1;
+}
+
+static void
+obj_ec_iom_merge(struct obj_reasb_req *reasb_req, uint32_t tgt_idx,
+		 const daos_iom_t *src, daos_iom_t *dst)
+{
+	struct daos_oclass_attr	*oca = reasb_req->orr_oca;
+	uint64_t		 stripe_rec_nr = obj_ec_stripe_rec_nr(oca);
+	uint64_t		 cell_rec_nr = obj_ec_cell_rec_nr(oca);
+	uint64_t		 end, rec_nr;
+	daos_recx_t		 hi, lo, recx, tmpr;
+	uint32_t		 i;
+	int			 rc;
+
+	D_ASSERT(tgt_idx < obj_ec_data_tgt_nr(oca));
+
+	D_SPIN_LOCK(&reasb_req->orr_spin);
```

Comment:
if client progressed by multiple threads, the different RPCs' completion callbacks possibly concurrently exuected.

Reply:
hmm, Do we support that? Is there use case on this?

Reply:
I think should support multi-threads client, and ever met issues for example dfuse/dfs client app.

Reply:
I can understand to add lock to protect the object on client stack. since object might be shared by multiple threads. But I do not understand to add lock to protect these reasb_req, and iom, why do we want to share them between multiple threads?

Reply:
one obj IO possibly involves multi-shards IO right? those shards IO possibly be completed concurrently, in shards IO's completion, need to merge each shard's replied IO map.

Reply:
yes, they can be completed concurrently. But they are in the same daos call, i.e. same io task, I do not see why we want the task/scheduler) shared by different threads. 

TBH, if we want to do that, there will be a lot part on client stack(especially shard layer) needs to be revisited IMHO.